tests/test_communication_manager.py

```diff
@@ -89,17 +89,28 @@ def test_communication_manager_can_initialize_and_train_and_average_model():
         RawEventTypes.NEW_SESSION.name,
         new_session_event
     )
-    while len(scheduler.processed) == 0:
+    timeout = time.time() + 3
+    while time.time() < timeout and len(scheduler.processed) == 0:
+        # initialization job
         scheduler.runners_run_next_jobs()
         time.sleep(0.1)
-    while len(scheduler.processed) == 1:
+    timeout = time.time() + 3
+    while time.time() < timeout and len(scheduler.processed) == 1:
+        # training job
         scheduler.runners_run_next_jobs()
         time.sleep(0.1)
     assert len(scheduler.processed) == 2
-    while len(scheduler.processed) == 2:
+    timeout = time.time() + 3
+    while time.time() < timeout and len(scheduler.processed) == 2:
+        # communication job
         scheduler.runners_run_next_jobs()
         time.sleep(0.1)
     assert len(scheduler.processed) == 3
+    # now the communication manager should be idle
+    scheduler.runners_run_next_jobs()
+    time.sleep(0.1)
+    assert len(scheduler.processed) == 3
```

Comment:
EDIT I see some assertions on length after some of the jobs, but you should still address having messages and explicit assertions for timeouts and asserting the length after the initialization job.

Also nitpick but tests could be stronger if you actually verify that the processed job matches what you expect it to be (i.e. tests would pass if three initialization jobs were run?)

Reply:
See latest commit

Reply:
Re: Runner errors, runner errors should be caught and logged so that should help the developer with debugging.

Reply:
You're right, but if the issue was originally Runner errors, then the outcome would still be a timeout (since failed jobs are discarded after three tries). 

So then I think you should clarify in the messages that the issue could be an actual timeout or Runner error (since it's currently impossible to determine which is which?)

Reply:
You can determine whether it's a timeout or Runner error by inspecting the log; if there are Runner errors and it keeps retrying, then it will show up in the log. If the test times out and there are no Runner errors, then it's a timeout. I've added this to the test docstring.

Reply:
I meant programmatically, so that assertion would be able to differentiate the actual error in its message.

Reply:
Resolving, we can discuss this further (if necessary) in the other thread